raft_node.py

```diff
@@ -0,0 +1,595 @@
+                pass
+
+            case FollowerState(receive_heartbeats_task=recv_heartbeats_task):
+                if recv_heartbeats_task is not None:
+                    # self._print('Heartbeat received. Cancelling old task')
+                    recv_heartbeats_task.cancel()
+
+                wait_for_heartbeat = asyncio.sleep(
+                    self.config.heartbeat_timeout_ms / 1000
+                )
+                self._state.receive_heartbeats_task = loop.create_task(
+                    wait_for_heartbeat
+                )
+
+                try:
+                    await self._state.receive_heartbeats_task
+                    self._state.receive_heartbeats_task = None
+                    self._print("Heartbeat timeout. Becoming a candidate")
+                    self._become_candidate()
+                except CancelledError:
+                    # Heartbeat received
+                    # self._print('Heartbeat task cancelled')
+                    pass
+
+    async def _hold_election(self: Self) -> None:
+        """Send VoteRequests to all other nodes. If in the middle
+        of the election we receive an RPC from a node with a
+        greater or equal term, we'll know another leader was
+        elected and we should instead become a follower.
+
+        If we don't win, we expect another node to win and
+        thus their next heartbeat should cancel the election.
+        If no other node wins, the election will time out and
+        we'll try again.
+        """
+        self._current_term += 1
+        self._voted_for = self.config.raft_server_port
+        self._print(f"Starting election for term {self._current_term}")
+
+        last_log_index, last_log_term = self._log.last_index_and_term()
+        votes: int = 1
+        self._print(
+            "Sending vote requests to "
+            f"{len(self._raft_rpc_clients.values())} clients"
+        )
+        vote_requests: Iterable[Coroutine[Any, Any, RequestVoteResponse]] = [
+            client.request_vote(
+                term=self._current_term,
+                candidate_id=self.config.raft_server_port,
+                last_log_index=last_log_index,
+                last_log_term=last_log_term,
+            )
+            for client in self._raft_rpc_clients.values()
+        ]
+
+        for coro in asyncio.as_completed(vote_requests):
+            try:
+                result: RequestVoteResponse = await coro
+                if result.vote_granted:
+                    votes += 1
+                    if votes > len(self.config.all_raft_server_ports) / 2:
+                        return
+            except AioRpcError:
+                # Can't reach one of the other servers; ignore. Either we'll
+                # win the election anyway or try again
+                pass
+
+        # If we didn't get a majority of votes, sleep ~forever and
+        # wait for either the new leader or the election timeout.
+        await asyncio.sleep(10000000)
```

Comment:
I guess this works! But maybe better to return a boolean out of this function, and then the caller can just check if it won the election or not (maybe you'll just end up moving the sleep into the child, but at least it'll be right next to the timeout, so the behavior should be more obvious)